rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,558 @@
+
+### Configuring a _custom field_
+
+The first step is to define how custom fields can be added/configured. A top to bottom approach (from more to less abstracted) will be followed to explain the configuration. Taking the following use case:
+
+_At CERN, I want to choose the experiment to which the record belong to, assuming experiments come from a controlled vocabulary. In addition, I want to be able to filter the record search by experiment._
+
+The instance administrator will change the configuraiton variables `RDM_RECORDS_CUSTOM_FIELDS` and `RDM_RECORDS_CUSTOM_FIELDS_UI` in the `invenio.cfg` configuration file. The following example shows only the configuration needed to **enable the custom field** in the data model not how to display them. This second part will be explained in the [_Adding-a-custom-UI-widget_](#Add a custom UI widget) section.
+
+```python
+RDM_RECORDS_CUSTOM_FIELDS = {
+    "experiment": VocabularyCF(
+        name="experiment",
+        vocabulary_id="cernexperiments",
+    ),
+}
+```
+
+> Note: In the above configuration and along this RFC configuration variables are prefixed with `RDM_RECORDS_`, this means that custom fields are being added to _records_. However, the same configuration variables are available in communities and will be used then adding custom fields to them. Namely, `COMMUNITIES_CUSTOM_FIELDS` and `COMMUNITIES_CUSTOM_FIELDS_UI`.
+
+All the available custom field types will be implemented in `<Type>CF` classes. These will be in charge of defining the properties that reflect the data model, such as Marshmallow schema and Elasticsearch mapping. In the previous example `VocabularyCF`.
+
+```python
+class BaseCF(ABC):
+"""Base Custom Field class."""
+
+    def __init__(self, name):
+        """Constructor."""
+        super().__init__()
+        self.name = name
+
+    @property
+    @abstractmethod
+    def mapping(self):
+        """Return the mapping."""
+        pass
+
+    @property
+    @abstractmethod
+    def schema(self):
+        """Marshmallow schema for custom fields."""
+        pass
+
+    @property
+    def ui_schema(self):
+        """Marshmallow UI schema for custom fields.
+
+        Controls how the field will be serialized in the UI when using the
+        `UIJSONSerializer`. If not overriden, the default implementation will use
+        the `BaseCF.schema` property.
+        """
+        return self.schema
+
+class TextCF(BaseCF):
+"""Text custom field."""
+
+    @property
+    def mapping(self):
+        return {"type": "text"}
+
+    @property
+    def schema(self):
+        return SanitizedUnicode()
+```
+
+Note that all custom fields will be added inside the `custom` parent field, at the same level than `metadata`.
+
+```json
+{
+   "metadata": {
+        ...
+    },
+    "custom_fields": {
+        "experiment": ...
```

Comment:
This should deal with namespacing as well:

```
"ext": {
    "dwc:kingdom": "....",
    "cern:experiment": "...."
}
```

Reply:
Namespacing would be dealt with complex objects, let's say "unresolved question" since complex objects are not supported right now. This would be something like:

```
"custom_fields": {
    "dwc": {
        "kingdom": "....",
    ",
    "cern": {
        "experiment": "....",
    }
}
```

I understand is not the previous implementation, but this way is more clear and compatible with the CF classes defined on the RFC. Do you see any drawbacks?